pkg/volume/util/selinux.go

```diff
@@ -196,3 +234,76 @@ func AddSELinuxMountOption(options []string, seLinuxContext string) []string {
 	// For example: dirsync,context="system_u:object_r:container_file_t:s0:c15,c25",noatime
 	return append(options, fmt.Sprintf("context=%q", seLinuxContext))
 }
+
+// SELinuxLabelInfo contains information about SELinux labels that should be used to mount a volume for a Pod.
+type SELinuxLabelInfo struct {
+	// SELinuxMountLabel is the SELinux label that should be used to mount the volume.
+	// The volume plugin supports SELinuxMount and the Pod did not opt out via SELinuxChangePolicy.
+	// Empty string otherwise.
+	SELinuxMountLabel string
+	// SELinuxProcessLabel is the SELinux label that will the container runtime use for the Pod.
+	// Regardless if the volume plugin supports SELinuxMount or the Pod opted out via SELinuxChangePolicy.
+	SELinuxProcessLabel string
+	// PluginSupportsSELinuxContextMount is true if the volume plugin supports SELinux mount.
+	PluginSupportsSELinuxContextMount bool
+}
+
+// GetMountSELinuxLabel returns SELinux labels that should be used to mount the given volume volumeSpec and podSecurityContext.
+// It does not evaluate the volume access mode! It's up to the caller to check SELinuxMount feature gate,
+// it may need to bump different metrics based on feature gates / access modes / label anyway.
+func GetMountSELinuxLabel(volumeSpec *volume.Spec, seLinuxContainerContexts []*v1.SELinuxOptions, podSecurityContext *v1.PodSecurityContext, volumePluginMgr *volume.VolumePluginMgr, seLinuxTranslator SELinuxLabelTranslator) (SELinuxLabelInfo, error) {
+	info := SELinuxLabelInfo{}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
+		return info, nil
+	}
+
+	if !seLinuxTranslator.SELinuxEnabled() {
+		return info, nil
+	}
+
+	pluginSupportsSELinuxContextMount, err := SupportsSELinuxContextMount(volumeSpec, volumePluginMgr)
+	if err != nil {
+		return info, err
+	}
+
+	info.PluginSupportsSELinuxContextMount = pluginSupportsSELinuxContextMount
+
+	// Collect all SELinux options from all containers that use this volume.
+	labels := sets.New[string]()
+	for _, containerContext := range seLinuxContainerContexts {
+		lbl, err := seLinuxTranslator.SELinuxOptionsToFileLabel(containerContext)
+		if err != nil {
+			fullErr := fmt.Errorf("failed to construct SELinux label from context %q: %w", containerContext, err)
+			return info, fullErr
+		}
+		labels.Insert(lbl)
+	}
+
+	// Ensure that all containers use the same SELinux label.
+	if labels.Len() > 1 {
+		// This volume is used with more than one SELinux label in the pod.
+		return info, &MultipleSELinuxLabelsError{labels: labels.UnsortedList()}
+	}
+	if labels.Len() == 0 {
```

Comment:
If no selinux context is specified for container, then shouldn't the selinux context specified in `PodSecurityContext` take effect?

Reply:
It does. `GetPodVolumeNames` called by the controller [here](https://github.com/kubernetes/kubernetes/pull/128242/files/34f741e94508605bb31898c48e53e0442a63c8ed#diff-3c16cee6572b0a942684edd7eaf53dcba87418529d3a85bd3fbe42081237feb9R429) calls `DetermineEffectiveSecurityContext` and that already mixes container's SecurityContext and PodSecurityContext together.

Reply:
So it does. Then this function parameter imo is incorrectly named and only reason `podSecurityContext` is supplied here because to determine change policy that needs to be applied. In which case - I would rename `seLinuxContainerContexts` to something like - `effectiveSeLinuxContext` and remove `podSecurityContext` from parameter list and supply `podSecurityContext.SELinuxChangePolicy` directly. But obviously this is not a blocker deal.

Reply:
Renamed to `effectiveSELinuxContainerLabels`. But I'd prefer to pass `PodSecurityContext`, because I don't need to copy `SELinuxChangePolicy` feature gate evaluation + check for `nil` when dereferencing PodSecurityContext.SELinuxChangePolicy to all callers.